### PR TITLE
Jax Cliffwalking Env

### PR DESCRIPTION
**Pull request by balisujohn**
# Description

This adds a `CliffWalkingJaxEnv` to the `tabular` environment suite, as well as the underlying `FuncEnv` subclass `CliffWalkingFunctional`. 

`CliffWalkingJaxEnv` is registered with the name `"Jax-CliffWalking-v0"` This PR still needs a bit of polish work, but you can try it out for yourself in GUI mode with 

````
python3 ./gymnasium/envs/tabular/cliffwalking.py
````

image

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: gymnasium/envs/tabular/cliffwalking.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+
+            shape = (4,12)
+            nS = 4 * 12
+                    # Cliff Location
+            cliff = np.zeros(shape, dtype=bool)
+            cliff[3, 1:-1] = True
+
+            
+            hikers = [
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/envs/tabular/cliffwalking.py:233]: Should this to line 275 not be done in `init_render_state`

File: gymnasium/envs/tabular/cliffwalking.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+            return render_state, np.transpose(np.array(pygame.surfarray.pixels3d(window_surface)),axes=(1, 0, 2))
+
+    def render_close(self, render_state: RenderStateType) -> None:
+        """Closes the render state."""
+        try:
+            import pygame
+        except ImportError as e:
+            raise DependencyNotInstalled(
+                "pygame is not installed, run `pip install gymnasium[classic_control]`"
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/envs/tabular/cliffwalking.py:304]: toy-text

Comment (balisujohn) [reply gymnasium/envs/tabular/cliffwalking.py]: fixed

File: gymnasium/envs/tabular/cliffwalking.py (hunk 1 above)
Comment (balisujohn) [reply gymnasium/envs/tabular/cliffwalking.py]: fixed

File: gymnasium/envs/tabular/cliffwalking.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    ) -> RenderStateType:
+        """Returns an initial render state."""
+        try:
+            import pygame
+        except ImportError:
+            raise DependencyNotInstalled(
+                "pygame is not installed, run `pip install gymnasium[classic_control]`"
+            )
+
+        rng = seeding.np_random(0)[0]
```
[the remark this answers is not in the document]
Comment (balisujohn) [reply gymnasium/envs/tabular/cliffwalking.py]: fixed

File: gymnasium/envs/tabular/cliffwalking.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 from gymnasium.wrappers import HumanRendering
 
 
-RenderStateType = Tuple["pygame.Surface"]  # type: ignore  # noqa: F821
+RenderStateType = Tuple["pygame.Surface", Tuple[int, int], int, Tuple[int, int], "numpy.ndarray", Tuple["pygame.Surface", "pygame.Surface", "pygame.Surface", "pygame.Surface"], "pygame.Surface", "pygame.Surface", Tuple[str, str], Tuple["pygame.surface", "pygame.surface"], Tuple[str, str], Tuple["pygame.surface", "pygame.surface"], "pygame.surface"]  # type: ignore  # noqa: F821
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/envs/tabular/cliffwalking.py:20]: Would we want the RenderStateType to be a NamedTuple?

Comment (balisujohn) [reply gymnasium/envs/tabular/cliffwalking.py]: changed

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
This looks good, and I think close to being merged. Thanks for your hard work.

I was able to get the type issue solved using `from __future__ import annotations` and `if TYPE_CHECKING: import pygame`
```python
from __future__ import annotations

from os import path
from typing import TYPE_CHECKING, NamedTuple

import jax
import jax.numpy as jnp
import numpy as np
from jax.random import PRNGKey

from gymnasium import spaces
from gymnasium.error import DependencyNotInstalled
from gymnasium.experimental.functional import ActType, FuncEnv, StateType
from gymnasium.experimental.functional_jax_env import FunctionalJaxEnv
from gymnasium.utils import EzPickle
from gymnasium.wrappers import HumanRendering


if TYPE_CHECKING:
    import pygame


class RenderStateType(NamedTuple):
    """A named tuple which contains the full render state of the Cliffwalking Env. This is static during the episode."""

    screen: pygame.surface
    shape: tuple[int, int]
    nS: int
    cell_size: tuple[int, int]
    cliff: np.ndarray
    elf_images: tuple[pygame.Surface, pygame.Surface, pygame.Surface, pygame.Surface]
    start_img: pygame.Surface
    goal_img: pygame.Surface
    bg_imgs: tuple[str, str]
    mountain_bg_img: tuple[pygame.Surface, pygame.Surface]
    near_cliff_imgs: tuple[str, str]
    near_cliff_img: tuple[pygame.Surface, pygame.Surface]
    cliff_img: pygame.Surface
```

**Comment by balisujohn** [comment]
added that typing fix

**Review by pseudo-rnd-thoughts — APPROVED** [review]

**Pull request merged**